activation/poet_client_test.go

```diff
@@ -82,6 +82,57 @@ func Test_HTTPPoetClient_Submit(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func Test_HTTPPoetClient_SubmitTillCtxDeadlineReached(t *testing.T) {
+	t.Parallel()
+
+	ctxWithDeadline, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+	defer cancel()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /v1/submit", func(w http.ResponseWriter, r *http.Request) {
+		now := time.Now()
+		deadline, _ := ctxWithDeadline.Deadline()
+		remain := deadline.Sub(now)
+
+		if remain.Seconds() <= time.Second.Seconds() {
+			resp, err := protojson.Marshal(&rpcapi.SubmitResponse{})
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			} else {
+				w.Write(resp)
+			}
+			return
+		}
+		http.Error(w, "some_error", http.StatusInternalServerError)
+	})
```

Comment:
This can be simplified, if the goal is to check if submit keeps retrying until the deadline.

```suggestion
	mux.HandleFunc("POST /v1/submit", func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "some_error", http.StatusInternalServerError)
	})
```

And then as @poszu said, assert that `Submit` returns `context.DeadlineExceeded` as error.

Reply:
Also maybe a question for @poszu, why do we pass the deadline both as `context.Context` and as `time.Time`? 

Would it make sense to extract the deadline for the request from the context? - This would then make it necessary to document this behaviour in the godoc of `Submit`.

Would it make sense to use the deadline passed to derive another context that has this as deadline (if a shorter deadline already exists that would be a no-op)

Also @poszu: the `deadline` value we pass to submit seems to be the following:

```go
	// we want to fetch the PoET proof latest 1 CycleGap before the publish epoch ends
	// so that a node that is setup correctly (i.e. can generate a PoST proof within the cycle gap)
	// has enough time left to generate a post proof and publish
	poetProofDeadline := publishEpochEnd.Add(-nb.poetCfg.CycleGap)
```

Which is 12 hours before the end of the epoch on mainnet, this seems non-sensical to me?

Reply:
This deadline is passed in the request to poet's `/submit` endpoint. The poet then checks if the currently open round (to which the node registers via /submit) will finish in time to provide the proof by this deadline. This mechanism solved the problem of some nodes registering in the wrong round (the next one), when they submitted the challenge close to the cycle gap window end. As the poet doesn't allow registering again in the same round with another challenge, these nodes were losing 2 epochs on that poet.